### DISCUSSION TITLE
How to limit the resources used by halfpipe (and specifically by ANTs)?

### DISCUSSION THREAD

**Discussion by leonardocerliani** [HALFpipe/HALFpipe#784] [General] [answered]
Is there a way to limit the number of threads which are used by each ANTs process when running halfpipe? 
In ANTs alone it can be done by setting the ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS environmental variable.
However if I try to set this before running halfpipe it is ignored.
Currently all the resources are used, and this is clogging our server.

**Comment by HippocampusGirl** [comment, marked as answer]
Hi @leonardocerliani,

This can be tuned using the command line flags `--nipype-omp-nthreads` and `--nipype-n-procs`. The first one will control the number of threads used for each command, for example by setting `ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS` etc., and the second one controls how many commands are run simultaneously.

For example, you can run HALFpipe like this to limit the usage to one core
```bash=
singularity run --bind /data halfpipe-1.2.3.sif --nipype-omp-nthreads 1 --nipype-n-procs 1
```

**Reply by leonardocerliani** [reply]
Thank you Lea @HippocampusGirl.
This works for singularity. I was wondering if there is something similar for docker. I managed to find a way to limit the number of cpus directly with a docker container option, but in case there is an option provided by halfpipe, it would be good to know.
